doc/reverse_engineering/formations.md

```diff
@@ -0,0 +1,289 @@
+* Formations will split on occasions where the obstruction is small (such as a single tree or a building). This was probably implemented to avoid formations to get stuck at a single choke point, e.g. a breach in the wall, when there are other optional paths nearby.
+* Units within the same formation don't collide with each other when the formation is established. It is assumed that the pathfinding algorithm of AoE2 was not capable enough to handle the complexity of the formation system.
+* Units do not know how far they have to travel and are following a rough path that is recalculated if they encounter an obstruction. Therefore units will sometimes transition from marching to combat formations far too early.
+
+## Algorithmic solutions
+
+The proposed algorithm has a run time of `O(n)`. Therefore, implementing it shouldn't affect performance that much, even for large formations over 40 units. `player_selection` is a set of units that the player selected and that is passed to the algorithm.
+
+### Sort units into subformations (Pseudocode)
+
+```
+INPUT:
+- Set of units: player_selection
+- Formation type: formation_type
+```
+```
+Initilization:
+
+# subformations for military units
+subformation[1] = new subformation();
+subformation[2] = new subformation();
+subformation[3] = new subformation();
+subformation[4] = new subformation();
+
+# dummy subformation for every unit that will
+# not form formations
+subformation[5] = new subformation();
+```
+```
+Algorithm:
+
+for unit u in player_selection do
+{
+    switch(u.GROUP_ID)
+    {
+        case 912:
+        case 947: subformation[1].add(u)
+        break();
+        case 906: subformation[2].add(u)
+        break();
+        case 900:
+        case 923:
+        case 936:
+        case 944:
+        case 955: subformation[3].add(u)
+        break();
+        case 902:
+        case 913:
+        case 918:
+        case 919:
+        case 920:
+        case 935:
+        case 943:
+        case 951:
+        case 959: subformation[4].add(u)
+        break();
+        case 922:
+        switch(u.LINE_ID)
+        {
+          case -294: subformation[1].add(u)
+          break();
+          case -293: subformation[2].add(u)
+          break();
+          case -283:
+          case -284:
+          case -292: subformation[3].add(u)
+          break();
+          case -285:
+          case 706: subformation[4].add(u)
+          break();
+          default: subformation[5].add(u)
+          break();
+        }
+        default: subformation[5].add(u)
+        break();
+    }
+}
+
+subformation[1].order_units(formation_type)
+subformation[2].order_units(formation_type)
+subformation[3].order_units(formation_type)
+subformation[4].order_units(formation_type)
+subformation[5].order_units(formation_type)
```

Comment:
shouldn't be that hard to come up with and algorithm. I'll have a look and write it all down..